From 9116d67579e7e1781b74bcf765d45944a94b2f62 Mon Sep 17 00:00:00 2001
From: Tim Yao <tim.yao@amlogic.com>
Date: Thu, 26 Mar 2015 14:59:26 -0700
Subject: [PATCH 5577/5965] PD#104752: h264: fix H.264 I only mode for
 interlaced source.

1. Fix I only mode for interlaced field source.
2. Fix playback rate control for interlaced field source.

Change-Id: I2b595778bbf28e42faab474d3ad0949940c08cad
---
 drivers/amlogic/amports/vh264.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/amlogic/amports/vh264.c b/drivers/amlogic/amports/vh264.c
index d088cab6ca2c..2c9f70673885 100755
--- a/drivers/amlogic/amports/vh264.c
+++ b/drivers/amlogic/amports/vh264.c
@@ -1256,7 +1256,7 @@ static void vh264_isr(void)
                     last_pts_remainder = 0;
                 }
 
-            } else {
+            } else if ((READ_VREG(AV_SCRATCH_F) & 2) == 0) {
                 pts = pts_inc_by_duration(&pts, &last_pts_remainder);
                 pts_valid = 1;
             }
@@ -1342,6 +1342,12 @@ static void vh264_isr(void)
                 } else {
                     vf->type = poc_sel ? VIDTYPE_INTERLACE_BOTTOM : VIDTYPE_INTERLACE_TOP;
                 }
+
+                if (READ_VREG(AV_SCRATCH_F) & 2) {
+                    // swap field type for I only mode
+                    vf->type ^= VIDTYPE_INTERLACE_TOP ^ VIDTYPE_INTERLACE_BOTTOM;
+                }
+
 #ifdef NV21
                 vf->type |= VIDTYPE_VIU_NV21;
 #endif
@@ -1354,6 +1360,7 @@ static void vh264_isr(void)
                 vf->pts_us64= (pts_valid) ? pts_us64 : 0;
                 vf->canvas0Addr = vf->canvas1Addr = spec2canvas(&buffer_spec[buffer_index]);
                 vfbuf_use[buffer_index]++;
+                p_last_vf = vf;
                 vf->ready_jiffies64=jiffies_64;
 
                 if ((error_recovery_mode_use & 2) && error) {
-- 
2.19.0

